### PR TITLE
Payment and orders

### PR DESCRIPTION
**Pull request by The-Blue-K**
Payment, ProductOrder and ServiceOrder Controllers.

### CODE REVIEW

File: PoSSapi/Controllers/PaymentController.cs [hunk 1]
```diff
@@ -27,64 +36,97 @@ public class PaymentController : GenericController<Payment>
...
-    * <param name="customerId" example="">Id of the customer whose recieving the invoice</param>
-    */
+    /** <summary>Gets all payments for specific customer</summary>
+      */
     [ProducesResponseType(StatusCodes.Status200OK)]
     [ProducesResponseType(StatusCodes.Status404NotFound)]
-    [HttpPost("{id}/invoice/{customerId}")]
-    public ActionResult SendInvoiceToCustomer(string id, string customerId)
+    [HttpGet("{id}/invoice/{customerId}")]
+    public ActionResult<Payment> GetCustomerPayment([FromQuery][Required] string? customerId)
```
Comment (MigleMic) [inline PoSSapi/Controllers/PaymentController.cs]: image

Why two customer id's are required?

File: PoSSapi/Controllers/ProductOrderController.cs [hunk 2]
```diff
@@ -52,59 +70,122 @@ protected class OrderProductReturnObject
...
     {
-        var productOrder = RandomGenerator.GenerateRandom<ProductOrder>(id);
+        var productOrder = _productOrderRepository.GetProductOrder(id);
+        
+        if (productOrder == null)
+        {
+            return NotFound();
+        }
+
         productOrder.OrderStatus = status;
```
Comment (MigleMic) [inline PoSSapi/Controllers/ProductOrderController.cs:95]: You don't update the database, so it doesn't actually change the status

File: PoSSapi/Controllers/ServiceOrderController.cs [hunk 3]
```diff
@@ -5,42 +5,60 @@
...
+        if (itemsPerPage <= 0)
+        {
+            return BadRequest("itemsPerPage must be greater than 0");
+        }
+        if (pageNum < 0)
+        {
+            return BadRequest("pageNum must be 0 or greater");
+        }
+
+        var orders = _orderRepository.GetAllOrders();
```
Comment (MigleMic) [inline PoSSapi/Controllers/ServiceOrderController.cs:47]: Shouldn't it be all service orders?

File: PoSSapi/Controllers/PaymentController.cs [hunk 4]
```diff
@@ -5,17 +5,26 @@
...
 using System.ComponentModel.DataAnnotations;
+using PoSSapi.Repositories;
+using PoSSapi.Dtos;
 
 namespace Controllers;
 
 [ApiController]
 [Route("[controller]")]
-public class PaymentController : GenericController<Payment>
+public class PaymentController : ControllerBase
```
Comment (martynasslav) [inline PoSSapi/Controllers/PaymentController.cs:15]: The generic controller was responsible for basic GET, POST, PUT, DELETE actions. When we started using Entity Framework, these actions became not so basic, they needed to be associated with specific database sets.

These actions are missing and need implementation:
- POST /Payment
- GET /Payment/{id}
- PUT /Payment/{id}
- DELETE /Payment/{id}

File: PoSSapi/Controllers/ProductOrderController.cs [hunk 5]
```diff
@@ -4,42 +4,60 @@
...
 using PoSSapi.Tools;
+using PoSSapi.Repositories;
+using PoSSapi.Dtos;
 
 namespace PoSSapi.Controllers;
 
 [ApiController]
 [Route("[controller]")]
-public class ProductOrderController : GenericController<ProductOrder>
+public class ProductOrderController : ControllerBase
```
Comment (martynasslav) [inline PoSSapi/Controllers/ProductOrderController.cs:14]: - POST /ProductOrder
- GET /ProductOrder/{id}
- PUT /ProductOrder/{id}
- DELETE /ProductOrder/{id}
- PATCH /ProductOrder/{id}

File: PoSSapi/Controllers/ServiceOrderController.cs [hunk 6]
```diff
@@ -5,42 +5,60 @@
...
 using PoSSapi.Classes;
+using PoSSapi.Repositories;
+using PoSSapi.Dtos;
 
 namespace PoSSapi.Controllers;
 
 [ApiController]
 [Route("[controller]")]
-public class ServiceOrderController : GenericController<Order>
+public class ServiceOrderController : ControllerBase
```
Comment (martynasslav) [inline PoSSapi/Controllers/ServiceOrderController.cs:15]: - POST /ServiceOrder
- GET /ServiceOrder/{id}
- PUT /ServiceOrder/{id}
- DELETE /ServiceOrder/{id}

File: PoSSapi/Controllers/PaymentController.cs (hunk 1 above)
Comment (martynasslav) [reply PoSSapi/Controllers/PaymentController.cs]: This action is implemented incorrectly. Based on the technical document, this action should only send an email to the customer (this is out-of-scope for us, really). Please review the technical document again.

Perhaps, compile the needed information into a field and let's pretend we send it – add a comment where that function would be called and take place.

**Review by martynasslav — CHANGES_REQUESTED** [review]

File: PoSSapi/Controllers/ProductOrderController.cs [hunk 7]
```diff
@@ -60,6 +60,21 @@ protected class OrderProductReturnObject
...
+    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(ProductOrder))]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    [HttpGet("{id}", Name = "GetProductOrder")]
+    public ActionResult<ProductOrder> GetProductOrder(string id)
+    {
+        var productOrder = _productOrderRepository.GetProductOrder(id);
+
+        if (productOrder == null)
+        {
+            return NoContent();
```
Comment (martynasslav) [inline PoSSapi/Controllers/ProductOrderController.cs]: It should be NotFound()

File: PoSSapi/Controllers/ServiceOrderController.cs [hunk 8]
```diff
@@ -61,6 +61,21 @@ protected class OrderServiceReturnObject
...
+    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(Order))]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    [HttpGet("{id}", Name = "GetServiceOrder")]
+    public ActionResult<Order> GetOrder(string id)
+    {
+        var order = _orderRepository.GetOrder(id);
+
+        if (order == null)
+        {
+            return NoContent();
```
Comment (martynasslav) [inline PoSSapi/Controllers/ServiceOrderController.cs]: NotFound()

**Review by martynasslav — CHANGES_REQUESTED** [review]

**Review by martynasslav — APPROVED** [review]

**Review by MigleMic — APPROVED** [review]

**Review by karoluzas — APPROVED** [review]

**Pull request merged**